src/main/java/com/yelp/nrtsearch/server/grpc/LuceneServer.java

```diff
@@ -495,25 +496,74 @@ public void startIndex(
         StreamObserver<AddDocumentResponse> responseObserver) {
 
       return new StreamObserver<AddDocumentRequest>() {
-        // TODO make this a config
-        private static final int MAX_BUFFER_LEN = 100;
-        private long count = 0;
-        Queue<AddDocumentRequest> addDocumentRequestQueue =
-            new ArrayBlockingQueue<>(MAX_BUFFER_LEN);
         List<Future<Long>> futures = new ArrayList<>();
+        // Map of {indexName: addDocumentRequestQueue}
+        Map<String, ArrayBlockingQueue<AddDocumentRequest>> addDocumentRequestQueueMap =
+            new ConcurrentHashMap<>();
+        // Map of {indexName: count}
+        Map<String, Long> countMap = new ConcurrentHashMap<>();
+        private static final int DEFAULT_MAX_BUFFER_LEN = 100;
+
+        private int getAddDocumentsMaxBufferLen(String indexName) {
+          try {
+            return globalState
+                .getIndex(indexName)
+                .getSaveState()
```

Comment:
@erikyangs sorry didn't look at this PR before you merged it. The `getSaveState` method is synchronized https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/IndexState.java#L636 and we are calling this for every batch of requests. Apart from being slow, this might also reintroduce the deadlocks we had before.  I was thinking about adding a getter for `addDocumentsMaxBufferLen` in `IndexState`, but that would also be slightly slow since the variable is volatile.
cc: @umeshdangat

Reply:
+1 on adding getter on `volatile int addDocumentsMaxBufferLen`.  It should be faster to read volatile variable than  read from JSONObject.